bvm/ballerina-core/src/main/java/org/ballerinalang/model/values/BStream.java

```diff
@@ -97,33 +103,36 @@ public BStructType getConstraintType() {
      *
      * @param data the data to publish to the stream
      */
-    public void publish(BStruct data) {
-        if (data.getType() != this.constraintType) {
-            throw new BallerinaException("incompatible types: object of type:" + data.getType().getName()
+    public void publish(BValue data) {
+        if (!data.getType().equals(this.constraintType) && !(constraintType instanceof BUnionType
```

Comment:
Is it possible to extract `data.getType()` since it is used in multiple places?

Reply:
Fixed in a02c67b